pkg/common/transaction.go

```diff
@@ -0,0 +1,38 @@
+package common
+
+import (
+	"github.com/nrc-no/core/pkg/store"
+	"gorm.io/gorm"
+)
+
+type TransactionManager interface {
+	Begin() *gorm.DB
+	Commit(tx *gorm.DB) *gorm.DB
+	Rollback(tx *gorm.DB) *gorm.DB
+}
+
+type transactionManager struct {
```

Comment:
Not sure about the use of `TransactionManager` and `transactionManager`... We already have these methods on the `db` object (either Gorm or SQLx or sql.DB)

Reply:
I see, I think its a bit early to introduce this since the `sql.DB` is exclusively called from whatever `store` we have. We kind of don't need yet the handlers to have access to this interface. I think it's likely that it will remain like that, basically that the handler will simply `Unmarshal` `Validate` `Call Store.Something`, where the `Store` will do pretty much everything related to DB calls and so on

Reply:
I think we definitely do need it. When you get to the service layer you'll see how.

```
Controller
  Unmarshal
  Validate
  Call Service.Create
Service
  TransactionManager.Begin
  Call Store.InsertA
  Call Store.InsertB
  Iterate over x:
     Call Store.InsertC
  TransactionManager.Commit
Controller
  Validate
  Marshal
```

Reply:
But I don't see why we need a service layer ? This microservice "business task" is to talk to the database. So there is no need to encapsulate the database layer in an `infrastructure layer` (if you think DDD design). The database concepts should be first-class citizen of this whole microservice. What exactly would be the role of the `service` layer in that case ? I think the `core-db-api` is a really fancy `store`

Reply:
I think regardless of what we call them there is a need for 3 layers (after the controller):

1. Layer A - Orchestrating multiple database calls (service)
2. Layer B - Executing a single database call (store)
3. Layer C - Building database statements (sql builder)

Reply:
I disagree. In all layers there is the same concept which is calling the database. These seem to me that they are just 3 distinct functions in the same layer rather than their own package with their own concepts. 
I think we do need some tooling to craft sql queries. But basically the problem seems to be

1. convert whatever payload to a series of operations
2. do sheanigans for reconciliation
3. use a DAG to sort the operations by dependencies
4. start transaction
5. perform each operation
6. end transaction
7. report

I really think we should stick to the most simple approach, since whatever we implement in this layer we basically have to fully and accurately replicate on the client side in Typescript. This part of the code should be able to run in client or server mode. If we distribute this in a N-Layered approach, then we need the same N-Layered approach for the client side as well. I honestly think we don't need it. It's going to be dense for sure, but this will be the logic for our core database engine for offline use as well.

Reply:
One option we could investigate would be a program that 
1. Accepts records as inputs
2. Outputs SQL statements, expects those statements to be executed

So basically
```
1. client send RECORDS to PROGRAM
2. PROGRAM outputs SQL
3. client executes SQL and sends results to PROGRAM
4. ... 
```

In this way, this program would simply accept inputs and outputs the necessary SQL statements that need to be executed to accurately reconcile the records.

```
# pseudocode

engine := newEngine("postgres")
engine.OnSQL(func (sqlStatement string) (interface{}, error){
  fmt.Println("Engine wants to execute statement:", sqlStatement)
  return db.Exec(sqlStatement)
}}
engine.PutRecords(records)
```

Reply:
Basically a program that can maintain our invariants by converting a set of RECORD INSTRUCTIONS to SQL

Reply:
Our `handlers` would basically just be exposing this engine over HTTP

Reply:
I get your point about the service essentially being the store, but I don't like this `engine` api.

> 1. convert whatever payload to a series of operations
> 2. do sheanigans for reconciliation
> 3. use a DAG to sort the operations by dependencies
To me this is the service layer. When creating a REST resource we are actually creating multiple database resources. By putting this in the store we have to replicate it for whatever db we use. This belongs in a different layer that interacts with the store through an abstract interface.

Something like this could take away the tx/db dependancy from the service layer.
```
func (d *entityService) Create(ctx context.Context, entity types.Entity) (*types.Entity, error) {
	storeInstance := store.NewInstance()

	storeInstance.AddStatements(store.InsertEntityDefinition())

	for i, attribute := range entity.Attributes {
		store.AddStatements(store.InsertAttribute())
	}

	for i, relationship := range entity.Relationships {
		store.AddStatements(store.InsertEntityRelationship())
	}

	store.ExecuteStatements()
}
``` 
`storeInstance` keeps a list of statements then `ExecuteStatements` runs them sequentially within a transaction.

We could move the `Service.Create` to `Store.Create` and make the other store methods private, but I think this is overloading the store.

I think having this layered approach is actually simpler. Each layer has a clearer responsibility and is easier to test. Layers or not, all of the logic stays the same and we'd have to replicate it anyway.

Reply:
> whatever we implement in this layer we basically have to fully and accurately replicate on the client side in Typescript
Build the api in typescript so we can reuse it?

Reply:
I would say yes in the Java, .Net, Python, Ruby, dot net or other programming paradigms where the frameworks encourage you to use dependency injection containers like Spring, Windsor & al., automatic field injection and so on. But the advantage and strength of go is that it provides enough out of the standard library to create an application that is almost as fast as raw C++, supports extreme parallelization, etc. without needing any third party library. It is MUCH faster than NodeJs, Python, and others. It is also much easier to reason about because there is no framework-specific, automagical, convention-over-configuration stuff. It's all there in the code. There's also no annotation-based logic, such as NestJS @Controller and others.

Another advantage is portability, since you can compile a go program into a portable executable that runs on all platforms, x86, windows, ios, raspberry pies, wasm, etc. You can end up with a container that contains a single file, the executable itself, which is a huge win for security, performance, and eventually cost. The recurring problem with NodeJS, python & al, is that you end up with thousands of dependencies, direct or indirect. And it makes it extremely difficult to bundle, and very long to build. By removing these dependencies, you also remove the need to keep up with vulns that are spread over thousands of libraries, and have to manage a very limited subset instead. NodeJS's founder even publicly said that he would rather use Go (or deno) for distributed apis rather than Node. 

Unless there is a library that absolutely requires it, I don't see any reason not to use go.

And in the microservice architecture we are leaning towards, the service is the program. I think there would need to be a very specific usecase if we were to use services within microservices. For example, in a monolith app, we would probably have some kind of a `CoreDbService` or something like that. But with microservices, we already made that cut upstream, and divided code so that `CoreDbService` is its own microservice. So it seems to me that introducing a N-Layered architecture within a microservice architecture is overkill.

What we are building here is not a service per-se. We're more like coming up with a protocol, and programming a node to be part of a network of peer-to-peer clients, so that the data is available at all times. I think that's the main shift in how we think about this project vs other regular web application projects. This is not a web application. Given a node in the network, it would see the address of this server like the address of any other probably

```
# peers.ts
{"peers": [10.22.104.203, 11.103.34.21, ...]}
```

Thie `core-db` resembles more a storage node than a web application. A node that only provides the Durability guarantee of a distributed, p2p storage solution. Which in a way, would perhaps make more sense if it was a Pull-based system rather than a Push-based system. That would probably make it much easier for the reconciliation logic, as each actor in the network would be responsible to keep its data up to date, rather than expecting other actors to push data to it. If each actor exposes a read api, it is sufficient for each actor to keep things up to date within themselves. In a way, there is no need for a public write api. Each actor could poll other actors in the network in a pull-based fashion, rather than having to deal with potential concurrent pushes to itself.

We could use something like the `gossip` protocol to distribute the latest timestamp of the latest change etc. between all actors in the network. Those could choose to pull the data when they see fit. If there is a discrepancy between the lastest local pulled state vs the gossiped latest state, then the actor knows it should pull stuff.

Reply:
Advantage of gossip is that it will support fragmented networks, such as when internet goes down, and is extremely fast

Reply:
I'm not really advocating for typescript as I agree with most of your points about go. The main benefit it would have is development speed and code sharing with the frontend. I think it's not the best choice, but probably the more pragmatic choice given a team of our size. Again, not suggesting we switch.

I don't agree with your point about not using dependency injection in go. Sure it may not explicitly encourage it, but it's still a good pattern to use. For example when we switch from writing to standard sql tables to writing to an event stream, we can easily swap out a single module.

I don't think a microservice architecture is a reason not to use a layered architecture within the service. If a service requires it then we should at it. Otherwise it can easily become very bloated and difficult to work with.

On the service layer, I think there are other use cases for it. In the new forms api it will need to communicate with the core-db api. I suspect this will warrant a service layer, although I haven't thought through this in detail yet.

On the distributed stuff. Yeah that all sounds good (without thinking too much on it), but I think we need to be pragmatic and consider this as a web service for now. If we start trying to design this as a p2p system without actually making a p2p system we'll end up down a rabbit hole and nothing will get done.

Reply:
I've pushed up an example to `models/foo` that removes the service layer. I don't think it's right as the logic in `Create` is generic enough to be removed from the postgres implementation. But I thought it's worth having an example there to discuss.